test/Rings-conformance-tests.jl

```diff
@@ -21,6 +21,148 @@ function equality(a::T, b::T) where T <: AbstractAlgebra.NCRingElement
    end
 end
 
+function test_mutating_op_like_zero(f::Function, f!::Function, A)
+  a = deepcopy(A)
+  a = f!(a)
+  @test equality(a, f(A))
+end
+
+function test_mutating_op_like_neg(f::Function, f!::Function, A, B)
+  a = deepcopy(A)
+  b = deepcopy(B)
+  a = f!(a, b)
+  @test equality(a, f(B))
+  @test b == B
+
+  a = deepcopy(A)
+  a = f!(a)
+  @test equality(a, f(A))
+end
+
+function test_mutating_op_like_divexact(f::Function, f!::Function, A, B, C)
+  a = deepcopy(A)
+  b = deepcopy(B)
+  c = deepcopy(C)
+  a = f!(a, b, c)
+  @test equality(a, f(B, C))
+  @test b == B
+  @test c == C
+
+  a = deepcopy(A)
+  b = deepcopy(B)
+  a = f!(a, a, b)
+  @test equality(a, f(A, B))
+  @test b == B
+
+  if typeof(A) == typeof(B)
+    a = deepcopy(A)
+    b = deepcopy(B)
+    a = f!(a, b, a)
+    @test equality(a, f(B, A))
+    @test b == B
+  end
```

Comment:
```suggestion
  if typeof(A) == typeof(B)
    a = deepcopy(A)
    b = deepcopy(B)
    a = f!(a, b, a)
    @test equality(a, f(B, A))
    @test b == B
  end
```
I don't think this is OK, just because the types are equal doesn't mean you can change the order. Think of `divexact(4,2)` vs. `divexact(2,4)`.

Instead I would suggest to check `f!(b, a, b)`. And if you do that, then perhaps there is no need for a separate `test_mutating_op_like_divexact` after all?

Reply:
Thinking some more along this vein, perhaps the arguments to `test_mutating_op_like_add` (and `test_mutating_op_like_divexact` if we retain it) should include just two values `a, b`. The third value really is only needed for storage purposes and presumably we can get that via deepcopy? (although, I am also worrying about relying on `deepcopy` too much here, but that's a separate concern... anyway `zero(parent(a))` might be a better source, dunno)

Next, I was wondering whether we should arrange things so that e.g. `b` is always the last argument, by using these input combos:
- zero(parent(a)), a, b
- a, a, b
- b, a, b  (only if a, b have same parent)
- a, b, b  (only if a, b have same parent)
- b, b, b  (only if a, b have same parent)
- a, b
- b, b  (only if a, b have same parent)

The idea being that this way the "divisor" always stays on the right side. And we are testing the parent of `a` (so if `b` has a different parent` we don't mix).

But then we loose the ability to test e.g. `add!(::ZZRingElem, ::Integer, ::ZZRingElem)` by passing an `Integer` for `a` and a `ZZRingElem` for `b`, because we implicitly assume the ring being tested is `parent(a)`. We could recover this if we passed the "target ring" `R` (= the parent of `a`), and then used these input combos (one may wish to also add a check that least one of `a` and `b` has `R` as its parent):
- zero(R), a, b
- a, b
- a, a, b  (only if R == parent(a))
- b, a, b  (only if R == parent(b))
- a, b, b  (only if R == parent(a) == parent(b))
- b, b, b  (only if R == parent(a) == parent(b))
- b, b  (only if R == parent(a) == parent(b))

Reply:
I like this and try to implement all of this now. One little thing: I would not only use `zero(R)` as the temporary result place, but something else as well, e.g. because for 0 we cannot distinguish `mul!` and `addmull!`